src/source-dpdk.c

```diff
@@ -0,0 +1,494 @@
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleDecodeDPDKRegister(void)
+{
+    tmm_modules[TMM_DECODEDPDK].name = "DecodeDPDK";
+    tmm_modules[TMM_DECODEDPDK].ThreadInit = DecodeDPDKThreadInit;
+    tmm_modules[TMM_DECODEDPDK].Func = DecodeDPDK;
+    tmm_modules[TMM_DECODEDPDK].ThreadExitPrintStats = NULL;
+    tmm_modules[TMM_DECODEDPDK].ThreadDeinit = DecodeDPDKThreadDeinit;
+    tmm_modules[TMM_DECODEDPDK].cap_flags = 0;
+    tmm_modules[TMM_DECODEDPDK].flags = TM_FLAG_DECODE_TM;
+}
+
+static inline void DPDKDumpCounters(DPDKThreadVars *ptv)
+{
+#ifdef PACKET_STATISTICS
+    // Prevents access to other fields of stats structure
+    if (unlikely(ptv->queue_id >= RTE_ETHDEV_QUEUE_STAT_CNTRS))
+        return;
+
+    struct rte_eth_stats eth_stats;
+    rte_eth_stats_get(ptv->port_id, &eth_stats);
+
+    uint64_t th_pkts = StatsGetLocalCounterValue(ptv->tv, ptv->capture_dpdk_packets);
+    uint64_t th_drops = StatsGetLocalCounterValue(ptv->tv, ptv->capture_dpdk_drops);
+
+    StatsAddUI64(ptv->tv, ptv->capture_dpdk_packets, eth_stats.q_ipackets[ptv->queue_id] - th_pkts);
+    StatsAddUI64(ptv->tv, ptv->capture_dpdk_drops, eth_stats.q_errors[ptv->queue_id] - th_drops);
+    SC_ATOMIC_ADD(ptv->livedev->pkts, eth_stats.q_ipackets[ptv->queue_id] - th_pkts);
+    SC_ATOMIC_ADD(ptv->livedev->drop, eth_stats.q_errors[ptv->queue_id] - th_drops);
+#endif
+}
+
+static void DPDKReleasePacket(Packet *p)
+{
+    int retval;
+    /* Need to be in copy mode and need to detect early release
+       where Ethernet header could not be set (and pseudo packet) */
+    if ((p->dpdk_v.copy_mode != DPDK_COPY_MODE_NONE) && !PKT_IS_PSEUDOPKT(p) &&
+            p->dpdk_v.copy_mode == DPDK_COPY_MODE_IPS && !PACKET_TEST_ACTION(p, ACTION_DROP)) {
+        retval =
+                rte_eth_tx_burst(p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+        if (unlikely(retval < 1)) {
+            rte_pktmbuf_free(p->dpdk_v.mbuf);
+            p->dpdk_v.mbuf = NULL;
+        }
+    } else {
+        rte_pktmbuf_free(p->dpdk_v.mbuf);
+        p->dpdk_v.mbuf = NULL;
+    }
+
+    PacketFreeOrRelease(p);
+}
+
+/**
+ *  \brief Main DPDK reading Loop function
+ */
+TmEcode ReceiveDPDKLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+    Packet *p;
+    uint16_t nb_rx;
+    time_t last_dump = 0;
+    time_t current_time;
+
+    DPDKThreadVars *ptv = (DPDKThreadVars *)data;
+    TmSlot *s = (TmSlot *)slot;
+
+    ptv->slot = s->slot_next;
+
+    PacketPoolWait();
+    while (1) {
+        if (unlikely(suricata_ctl_flags != 0)) {
+            SCLogInfo("Stopping Suricata!");
+            DPDKDumpCounters(ptv);
+            break;
+        }
+
+        nb_rx = rte_eth_rx_burst(ptv->port_id, ptv->queue_id, ptv->received_mbufs, BURST_SIZE);
+        if (unlikely(nb_rx == 0)) {
+            continue;
+        }
+
+        ptv->pkts += (uint64_t)nb_rx;
+        for (int i = 0; i < nb_rx; i++) {
```

Comment:
i is declared as int and nb_rx is declared as unit16_t,. minor nit pick use proper type cast. or rework variable `i`